### PR TITLE
New migration generator proof-of-concept

### PR DESCRIPTION
**Pull request by codyebberson**
Prototype for internal review

### What is this?

A new database migration generator that works in 3 steps:

1. Build an in-memory representation of the current database ("start")
2. Build an in-memory representation of the desired database from FHIR spec ("target")
3. Table by table, column by column, generate the SQL statements to move from "start" to "target"

The previous database migration generator was much less sophisticated.  For every migration, it required some manual tuning to output the desired migration.

This new migration generator is intended to be fully automated.

### Why do this?

Short term:

1. The new FHIRcast implementation would benefit from the FHIR R4B type `SubscriptionTopic`
2. I prototyped swapping in R4B (see https://github.com/medplum/medplum/tree/cody-r4b), which looks great
3. But to fully deploy R4B will require some heavy lifting from the migration generator

Longer term:

1. This could code move from `packages/generator` to `packges/server`
2. The server could run this automatically on first run to initialize a database
3. The server could run this automatically on new resource types / new search parameters
4. This would open up custom search parameters / custom indexes

### CODE REVIEW

File: packages/generator/src/migrate.ts [hunk 1]
```diff
@@ -258,63 +325,169 @@ function getColumnType(details: SearchParameterDetails): string {
...
+}
+
+function migrateColumns(b: FileBuilder, startTable: TableDefinition, targetTable: TableDefinition): void {
+  for (const targetColumn of targetTable.columns) {
+    const startColumn = startTable.columns.find((c) => c.name === targetColumn.name);
+    if (!startColumn) {
+      writeAddColumn(b, targetTable, targetColumn);
+    } else if (
+      startColumn.type.replaceAll('TIMESTAMP WITH TIME ZONE', 'TIMESTAMPTZ').trim() !==
+      targetColumn.type.replaceAll(' PRIMARY KEY', '').replaceAll(' DEFAULT FALSE', '').trim()
```
Comment (mattwiller) [inline packages/generator/src/migrate.ts:434]: What's going on here?

File: packages/generator/src/migrate.ts [hunk 2]
```diff
@@ -258,63 +325,169 @@ function getColumnType(details: SearchParameterDetails): string {
...
+  result += '_idx ON "';
+  result += tableName;
+  result += '"';
+
+  if (index.gin) {
+    result += ' USING gin ';
+  }
+
+  result += '(';
+  result += index.columns.map((c) => '"' + c + '"').join(', ');
```
Comment (mattwiller) [inline packages/generator/src/migrate.ts:486]:
```suggestion
  result += index.columns.map((c) => `"${c}"`).join(', ');
```

File: packages/generator/src/migrate.ts [hunk 3]
```diff
@@ -258,63 +325,169 @@ function getColumnType(details: SearchParameterDetails): string {
...
+  result += index.columns.join('_');
+  result += '_idx ON "';
+  result += tableName;
+  result += '"';
+
+  if (index.gin) {
+    result += ' USING gin ';
+  }
+
+  result += '(';
```
Comment (mattwiller) [inline packages/generator/src/migrate.ts:485]:
```suggestion
  result += '" ';

  if (index.gin) {
    result += 'USING gin ';
  }

  result += '(';
```

**Review by mattwiller — APPROVED** [review]
Love the approach: it's great to have this package getting into a robust state

File: packages/generator/src/migrate.ts [hunk 4]
```diff
@@ -12,8 +12,30 @@ import { readJson } from '@medplum/definitions';
...
+}
+
+interface ColumnDefinition {
+  name: string;
+  type: string;
+}
+
+interface IndexDefinition {
+  columns: string[];
+  gin?: boolean;
```
[the remark this answers is not in the document]
Comment (codyebberson) [reply packages/generator/src/migrate.ts:35]: Nice, like it.

File: packages/generator/src/migrate.ts (hunk 2 above)
Comment (codyebberson) [reply packages/generator/src/migrate.ts:486]: Agree

File: packages/generator/src/migrate.ts (hunk 1 above)
Comment (codyebberson) [reply packages/generator/src/migrate.ts:434]: Yeah...  This is a mismatch between "column definition" and "index definition"

I agree with your meta point, which is "this is confusing".... I'll move this logic to a helper function called "normalizeColumnDefinition" or something.

**Pull request merged**